packages/adminrouter/extra/src/test-harness/tests/test_service.py

```diff
@@ -553,3 +554,198 @@ def test_if_scheme_is_honoured_in_mesos_scheduler_entry(
+            assert resp.status_code == 200
+            data = resp.json()
+            assert data['endpoint_id'] == 'http://127.0.0.1:16000'
+
+            # If AR wait for cache during resolve, then time spent should be
+            # greater than the stall time that has been set. Due to the fact
+            # that update coroutines are not separated yet, this will be
+            # slightly higher than: 2 * (backend_request_timeout * 0.5)
+            assert t_spent > backend_request_timeout * 0.5
+
+    def test_if_ar_with_empty_cache_waits_for_mesos_during_service_resolve(
+            self, mocker, nginx_class, valid_user_header):
+        # Make service endpoint resolve only Marathon-related data:
+        mocker.send_command(
+            endpoint_id='http://127.0.0.1:8080',
+            func_name='set_apps_response',
+            aux_data={"apps": []})
+        mocker.send_command(
+            endpoint_id='http://127.0.0.1:8123',
+            func_name='set_srv_response',
+            aux_data=EMPTY_SRV)
+
+        # Make Mock endpoint stall a little, make sure AR cache update timeouts
+        # are big enough to swallow it:
+        backend_request_timeout = 6
+        refresh_lock_timeout = backend_request_timeout * 2
+
+        # Make period cache refreshes so rare that they do not get into
+        # picture:
+        ar = nginx_class(cache_first_poll_delay=1200,
+                         cache_poll_period=1200,
+                         cache_expiration=1200,
+                         cache_max_age_soft_limit=1200,
+                         cache_max_age_hard_limit=1800,
+                         cache_backend_request_timeout=backend_request_timeout,
+                         cache_refresh_lock_timeout=refresh_lock_timeout,
+                         )
+        url = ar.make_url_from_path("/service/scheduler-alwaysthere/")
+
+        mocker.send_command(endpoint_id='http://127.0.0.2:5050',
+                            func_name='always_stall',
+                            aux_data=backend_request_timeout * 0.5)
+
+        # Measure the time it took and the results:
+        with GuardedSubprocess(ar):
+            t_start = time.time()
+            resp = requests.get(url,
+                                allow_redirects=False,
+                                headers=valid_user_header)
+
+            t_spent = time.time() - t_start
+            assert resp.status_code == 200
+            data = resp.json()
+            assert data['endpoint_id'] == 'http://127.0.0.1:16000'
+
+            # If AR wait for cache during resolve, then time spent should be
+            # greater than the stall time that has been set. Due to the fact
+            # that update coroutines are not separated yet, this will be
+            # slightly higher than: 2 * (backend_request_timeout * 0.5)
+            assert t_spent > backend_request_timeout * 0.5
+
+    def test_if_marathon_borkerage_prevents_resolving_via_mesos_state_summary(
```

Comment:
> Borkerage is not a word. Perhaps a better term would be test_if_broken_marathon.... Optional partly because it is not new to this codebase.

Sure, fixing.

> Personally I prefer to organise tests into more granular classes, and would maybe go for 
This is more complext than that - e.g. AR fixtures used across test harness may have per-module, per-class or per-test scope. You *could* organise some of the classes here, but others (esp. the ones that use per-class AR fixture) would either have to be kept as-is, or we would have to accept longer tests execution times.